### PR TITLE
Fixed #32740 -- Caught OSError from colorama.init().

### PR DESCRIPTION
**Pull request by carltongibson**
As per comment on ticket-32740 I'm not quite sure how (or indeed whether) to test this, so any thoughts (that aren't too horrible 😀) welcome.

### CODE REVIEW

**Comment by claudep** [comment]
Meh… as the reporter told, I'm not sure it's a good idea to init at load time anyway…

**Comment by carltongibson** [comment]
We had it nested initially... then it got moved out... I need to look back to what the comment was there. 🤔

**Comment by felixxm** [comment]
> We had it nested initially... then it got moved out... I need to look back to what the comment was there. thinking

Mea culpa, see https://github.com/django/django/pull/12387#discussion_r472872001

**Comment by carltongibson** [comment]
Thanks for finding that link @felixxm — I think this might be one of those _Style_ questions. 😄

**Comment by carltongibson** [comment]
Hmmm. 🤔

I'm having some difficulties getting colorama to behave here. It works fine at the top level, but as soon as I nest the `init()` call (inside `supports_color()`) it stops working. With the old CMD, the output is wrapped in what I'm taking to be escape codes, but they're not colored (no interpreted.) 

@MinchinWeb I don't suppose you have half a cycle to comment here do you? (You had colorama working correctly nested before yes?) I'm bemused. It looks a bit like your https://github.com/tartley/colorama/issues/233 but I can't work out why the nested makes a difference — maybe it's called too late... 🤔


Due to not handling the OSError from mod_wsgi, this bit from the colorama readme isn't quite true: 

> On other platforms, calling init() has no effect ... By design, this permits applications to call init() unconditionally on all platforms, after which ANSI output should just work.

But it should be.

**Comment by MinchinWeb** [comment]
Hey @carltongibson ! 

As a side, thanks for tackling these problems as they come up, and being so prompt on it. My head is starting to spin just trying to track of everything going on here!

---

Some comments:

> There's no reason to have colorama installed at all if you're not on Windows. ([source](https://code.djangoproject.com/ticket/32740#comment:5))

*colorama* also has some helper constants for the actual ANSI color codes. I.e. you can use `colorama.Fore.GREEN` for green text, rather than trying to figure out if you've escaped your ANSI code correctly.  These constants are usable on all platforms. So it's not entirely useless to have colorama installed on platforms other than Windows.

---

Re *mod_wsgi* raising `IOError` (cf. https://github.com/GrahamDumpleton/mod_wsgi/issues/675): As *mod_wsgi* effectively turns stdout into a "read only file", this is likely the "proper" error to raise. But to treat stdout as a file (or file-like) is very much the POSIX philosophy of "treating everything as a file"; on Windows, I'd expect it to be a COM object or something similiar. 

So coming from Windows, it's not intuitive that calling the (presumably) standard API on an object would return an error saying "this file is read-only".

Presumably, asking if a stream is closed should be possible without actually writing to it, so my preferred solution would be the *mod_wsgi* maintain that API endpoint. In theory, a program that's told that a stream is closed shouldn't be trying to write to it.

A patch to colorama to handle this IOError wouldn't be inappropriate either. [1]

From a practical standpoint though, it's probably fastest to patch Django.

---

> (I wondered if it were a multiple call issue with init, but ensure it's only called once had the same effect — it's nesting it that caused the issue.)

My understanding is that calling \_\_init\_\_ twice shouldn't be a problem; I suspect many deployments already do this. But practical tests beat theoretical understanding most times.

> You had colorama working correctly nested before yes?

Yes...? I don't recall specifically, but I wouldn't have let it sit if it wasn't working for me.

---

Re tartley/colorama#233: I'd forgotten about that. For whatever reason, I can't reproduce it today (on Python 3.8 on PowerShell 7.1.2), so maybe it got fixed by a 3.8 bugfix release? But I turned on my colored prompt now and it works, so that makes me happy!

---

As for the actual patch, the current implementation is basically catching the same as a bare `except` clause (by catching `BaseException`), which is generally unwanted. I think I would write the code thus (effectively combining the two try-except's):

~~~python
try:
    import colorama
    colorama.init()
except ImportError, IOError:
    HAS_COLORAMA = False
else:
    HAS_COLORAMA = True
~~~

---

[1]: if you ever want/need an example of an open source package that is much used, but thinly maintained, *colorama* is a good candidate. It's GitHub page says it's used 185,000 times and probably by everyone who runs Python on Windows; it's even vendored by [pip](https://github.com/pypa/pip/tree/main/src/pip/_vendor/colorama). However, it appears that it [only earns](https://tidelift.com/lifter/search/pypi/colorama) ~$50/month in donations.

I think the generally slow pace of releases reflects that: a huge install base so a desire not to break things, plus one guy doing it in his spare time.

---

I hope some of this rambling is helpful.

**Comment by carltongibson** [comment]
Hi @leon-matthews — I've pushed an adjustment (along the lines suggested by @MinchinWeb) — there's no reason for it not to be, but if you fancied giving that a test run, that would be great! 

@MinchinWeb — thanks for your comments yes. All to the good. 😉

File: django/core/management/color.py [hunk 1]
```diff
@@ -10,11 +10,11 @@
...
 try:
     import colorama
-except ImportError:
+    colorama.init()
+except (ImportError, OSError):
     HAS_COLORAMA = False
 else:
-    colorama.init()
-    HAS_COLORAMA = True
+    HAS_COLORAMA = False
```
Comment (felixxm) [inline django/core/management/color.py:17]:
```suggestion
    HAS_COLORAMA = True
```

Comment (carltongibson) [reply django/core/management/color.py]: Doh!

**Comment by carltongibson** [comment]
Squashed and Rebased.

File: docs/releases/3.2.4.txt [hunk 2]
```diff
@@ -15,3 +15,7 @@ Bugfixes
 
 * Fixed a bug in Django 3.2 where a system check would crash on an abstract
   model (:ticket:`32733`).
+
+* Fixed a crash in Django 3.2 that could occur when running ``mod_wsgi`` with
+  the recommended settings whilst the Windows ``colorama`` library was
```
Comment (felixxm) [inline docs/releases/3.2.4.txt:20]:
```suggestion
  the recommended settings while the Windows ``colorama`` library was
```
see #12976 and dbdc192ca30b70a9f22c6bc4b3155bbccd01c323 :wink:

**Review by felixxm — APPROVED** [review]
@carltongibson Thanks :+1:

File: docs/releases/3.2.4.txt (hunk 2 above)
Comment (carltongibson) [reply docs/releases/3.2.4.txt:20]: > The lyf so short, the craft so long to lerne.

😀

**Pull request merged**